src/SourceCode.Chasm/TreeNodeList.cs

```diff
@@ -393,8 +268,76 @@ public bool TryGetValue(string key, NodeKind kind, out TreeNode value)
 
         #region Helpers
 
-        private static ArgumentException CreateDuplicateException(TreeNode node)
-            => new ArgumentException($"Duplicate {nameof(TreeNode)} arguments passed to {nameof(TreeNodeList)}: ({node})");
+        private static ReadOnlyMemory<TreeNode> DistinctSort(TreeNode[] array, bool alreadyCopied)
+        {
+            Debug.Assert(array != null); // Already checked at callsites
+
+            // Optimize for common cases 0, 1, 2, N
+            switch (array.Length)
+            {
+                case 1: return array;
+
+                case 2:
```

Comment:
Always copy. The callsite may change the array after creating the TreeNodeList.

Reply:
Unless it comes from within TreeNodeList.

Reply:
Good catch